File: курсовая/c#/SalesReport.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text;

namespace Курсовая_работа_Музыкальный_магазин_с_
{
    public class SalesReport
    {
        private const int MAX_TOP_PRODUCTS = 10;      // максимальное количество товаров в топе продаж
        private const int MAX_ANALYSIS_ITEMS = 20;    // максимальное количество пунктов анализа

        private string period;                        // период отчета
        private double money;                         // общая выручка за период
        private int soldProducts;                     // общее количество проданных товаров

        private string[] topSales;                    // массив самых продаваемых товаров
        private int topSalesCount;                    // текущее количество товаров в топе

        private string[] analysis;                    // массив аналитических записей
        private int analysisCount;                    // текущее количество записей анализа

        // Конструктор по умолчанию
        public SalesReport()
        {
            money = 0.0;
            soldProducts = 0;
            topSalesCount = 0;
            analysisCount = 0;
            period = GetCurrentDate();
            AllocateMemory();
        }

        // Конструктор с параметрами
        public SalesReport(string reportPeriod)
        {
            period = reportPeriod;
            money = 0.0;
            soldProducts = 0;
            topSalesCount = 0;
            analysisCount = 0;
            AllocateMemory();
        }

        // Конструктор копирования (глубокое копирование)
        public SalesReport(SalesReport other)
        {
            period = other.period;
            money = other.money;
            soldProducts = other.soldProducts;
            topSalesCount = other.topSalesCount;
            analysisCount = other.analysisCount;

            // Выделяем новую память
            AllocateMemory();

            // Копируем содержимое массивов
            CopyArrays(other);
        }

        // Метод для поверхностного клонирования
        public SalesReport ShallowClone()
        {
            SalesReport clone = new SalesReport();

            // Копируем простые поля
            clone.period = this.period + " (shallow clone)";
            clone.money = this.money;
            clone.soldProducts = this.soldProducts;
            clone.topSalesCount = this.topSalesCount;
            clone.analysisCount = this.analysisCount;

            // ПОВЕРХНОСТНОЕ КОПИРОВАНИЕ: копируем ссылки на массивы
            clone.topSales = this.topSales;
            clone.analysis = this.analysis;

            return clone;
        }

        // Метод для глубокого клонирования
        public SalesReport DeepClone()
        {
            SalesReport clone = new SalesReport();

            // Копируем простые поля
            clone.period = this.period + " (deep clone)";
            clone.money = this.money;
            clone.soldProducts = this.soldProducts;
            clone.topSalesCount = this.topSalesCount;
            clone.analysisCount = this.analysisCount;

            // Глубокое копирование: копируем содержимое массивов
            for (int i = 0; i < topSalesCount; i++)
            {
                clone.topSales[i] = this.topSales[i];
            }

            for (int i = 0; i < analysisCount; i++)
            {
                clone.analysis[i] = this.analysis[i];
            }

            return clone;
        }

        // Вывод информации о отчете
        public void DisplayReportInfo()
        {
            Console.WriteLine("\n=== ИНФОРМАЦИЯ ОБ ОТЧЕТЕ ===");
            Console.WriteLine($"Период: {period}");
            Console.WriteLine($"Выручка: {money} руб.");
            Console.WriteLine($"Проданные товары: {soldProducts} шт.");
            Console.WriteLine($"Топ товаров: {topSalesCount}");
            Console.WriteLine($"Записей анализа: {analysisCount}");
        }

        // Модификация для демонстрации
        public void ModifyReportForDemo()
        {
            Console.WriteLine("\n=== МОДИФИКАЦИЯ ОТЧЕТА ДЛЯ ДЕМОНСТРАЦИИ ===");

            if (topSalesCount > 0)
            {
                topSales[0] = "МОДИФИЦИРОВАННЫЙ ТОВАР";
                Console.WriteLine($"Первый товар в топе изменен на: {topSales[0]}");
            }

            money += 1000.0;
            Console.WriteLine("Выручка увеличена на 1000 руб.");
        }

        // Выделение памяти для массивов
        private void AllocateMemory()
        {
            topSales = new string[MAX_TOP_PRODUCTS];
            analysis = new string[MAX_ANALYSIS_ITEMS];
        }

        // Копирование массивов (глубокое копирование)
        private void CopyArrays(SalesReport other)
        {
            // Копируем топ товаров
            for (int i = 0; i < other.topSalesCount; i++)
            {
                topSales[i] = other.topSales[i];
            }

            // Копируем анализ
            for (int i = 0; i < other.analysisCount; i++)
            {
                analysis[i] = other.analysis[i];
            }
        }

        // Основные методы
        public void GetReport()
        {
            Console.WriteLine("\n=== ОФОРМЛЕНИЕ ОТЧЕТА ПО ПРОДАЖАМ ===");
            Console.WriteLine($"Период отчета: {period}");
            InputReportData();
            SaveToFile();
            Console.WriteLine("Отчет успешно оформлен и сохранен!");
        }

        public void RemakeReport()
        {
            Console.WriteLine("\n=== ИЗМЕНЕНИЕ ОТЧЕТА ПО ПРОДАЖАМ ===");
            string filename = "sales_report_" + period + ".txt";

            if (!LoadFromFile(filename))
            {
                Console.WriteLine($"Отчет за период '{period}' не найден!");
                return;
            }

            Console.WriteLine("Текущие данные отчета:");
            DisplayReport();
            Console.WriteLine("\nВведите новые данные:");
            InputReportData();
            SaveToFile();
            Console.WriteLine("Отчет успешно изменен!");
        }

        private void InputReportData()
        {
            // Ввод выручки
            while (true)
            {
                Console.Write("Введите выручку: ");
                if (double.TryParse(Console.ReadLine(), out money) && money >= 0)
                {
                    break;
                }
                Console.WriteLine("Ошибка! Введите положительное число.");
            }

            // Ввод количества проданных товаров
            while (true)
            {
                Console.Write("Введите количество проданных товаров: ");
                if (int.TryParse(Console.ReadLine(), out soldProducts) && soldProducts >= 0)
                {
                    break;
                }
                Console.WriteLine("Ошибка! Введите целое положительное число.");
            }

            // Ввод топа товаров
            Console.WriteLine("\n=== ТОП ТОВАРОВ ПО ПРОДАЖАМ ===");
            topSalesCount = 0;

            for (int i = 0; i < MAX_TOP_PRODUCTS; i++)
            {
                Console.Write($"Введите товар №{i + 1} (или 'stop' для завершения): ");
                string product = Console.ReadLine();

                if (product == "stop" || string.IsNullOrEmpty(product))
                {
                    break;
                }

                topSales[topSalesCount++] = product;
            }

            // Ввод аналитики
            Console.WriteLine("\n=== АНАЛИТИКА ПРОДАЖ ===");
            analysisCount = 0;

            for (int i = 0; i < MAX_ANALYSIS_ITEMS; i++)
            {
                Console.Write($"Введите пункт аналитики №{i + 1} (или 'stop' для завершения): ");
                string analysisItem = Console.ReadLine();

                if (analysisItem == "stop" || string.IsNullOrEmpty(analysisItem))
                {
                    break;
                }

                analysis[analysisCount++] = analysisItem;
            }
        }

        public void SaveToFile()
        {
            string filename = "sales_report_" + period + ".txt";

            using (StreamWriter file = new StreamWriter(filename))
            {
                file.WriteLine(period);
                file.WriteLine(money);
                file.WriteLine(soldProducts);
                file.WriteLine(topSalesCount);

                for (int i = 0; i < topSalesCount; i++)
                {
                    file.WriteLine(topSales[i]);
                }

                file.WriteLine(analysisCount);

                for (int i = 0; i < analysisCount; i++)
                {
                    file.WriteLine(analysis[i]);
                }
            }
        }

        public bool LoadFromFile(string filename)
        {
            if (!File.Exists(filename))
            {
                return false;
            }

            string[] lines = File.ReadAllLines(filename);
            if (lines.Length < 5) return false;

            period = lines[0];
            money = double.Parse(lines[1]);
            soldProducts = int.Parse(lines[2]);
            topSalesCount = int.Parse(lines[3]);

            int index = 4;
            for (int i = 0; i < topSalesCount; i++)
            {
                if (index < lines.Length)
                {
                    topSales[i] = lines[index++];
                }
            }

            if (index < lines.Length)
            {
                analysisCount = int.Parse(lines[index++]);
                for (int i = 0; i < analysisCount; i++)
                {
                    if (index < lines.Length)
                    {
                        analysis[i] = lines[index++];
                    }
                }
            }

            return true;
        }

        public void DisplayReport()
        {
            Console.WriteLine("\n=== ОТЧЕТ ПО ПРОДАЖАМ ===");
            Console.WriteLine($"Период: {period}");
            Console.WriteLine($"Выручка: {money:F2} руб.");
            Console.WriteLine($"Проданные товары: {soldProducts} шт.");

            if (topSalesCount > 0)
            {
                Console.WriteLine("\nТоп товаров по продажам:");
                for (int i = 0; i < topSalesCount; i++)
                {
                    Console.WriteLine($"{i + 1}. {topSales[i]}");
                }
            }

            if (analysisCount > 0)
            {
                Console.WriteLine("\nАналитика продаж:");
                for (int i = 0; i < analysisCount; i++)
                {
                    Console.WriteLine($"- {analysis[i]}");
                }
            }

            Console.WriteLine("==========================");
        }

        private string GetCurrentDate()
        {
            return DateTime.Now.ToString("dd.MM.yyyy");
        }

        // Геттеры
        public string GetPeriod() { return period; }
        public double GetMoney() { return money; }
        public int GetSoldProducts() { return soldProducts; }
        public int GetTopSalesCount() { return topSalesCount; }
        public int GetAnalysisCount() { return analysisCount; }
    }
}

File: курсовая/c++/cpp/SalesReport.cpp
#define _CRT_SECURE_NO_WARNINGS
#include "SalesReport.h"
#include <iostream>
#include <iomanip>
#include <limits>
#include <sstream>
#include <cstring>

using namespace std;

// Конструктор по умолчанию
SalesReport::SalesReport() :
    money(0.0),
    soldProducts(0),
    topSalesCount(0),
    analysisCount(0)
{
    period = getCurrentDate();
    allocateMemory();

}

// Конструктор с параметрами
SalesReport::SalesReport(const std::string& reportPeriod) :
    period(reportPeriod),
    money(0.0),
    soldProducts(0),
    topSalesCount(0),
    analysisCount(0)
{
    allocateMemory();

}

// Конструктор копирования (глубокое копирование)
SalesReport::SalesReport(const SalesReport& other) :
    period(other.period),
    money(other.money),
    soldProducts(other.soldProducts),
    topSalesCount(other.topSalesCount),
    analysisCount(other.analysisCount)
{


    // Выделяем новую память
    allocateMemory();

    // Копируем содержимое массивов
    copyArrays(other);
}

// Оператор присваивания (глубокое копирование)
SalesReport& SalesReport::operator=(const SalesReport& other) {


    if (this != &other) {
        // Освобождаем старую память
        deallocateMemory();

        // Копируем простые поля
        period = other.period;
        money = other.money;
        soldProducts = other.soldProducts;
        topSalesCount = other.topSalesCount;
        analysisCount = other.analysisCount;

        // Выделяем новую память
        allocateMemory();

        // Копируем содержимое массивов
        copyArrays(other);
    }
    return *this;
}

// Деструктор
SalesReport::~SalesReport() {

    deallocateMemory();
}

// Метод для поверхностного клонирования
SalesReport* SalesReport::shallowClone() const {

    SalesReport* clone = new SalesReport();

    // Копируем простые поля
    clone->period = this->period + " (shallow clone)";
    clone->money = this->money;
    clone->soldProducts = this->soldProducts;
    clone->topSalesCount = this->topSalesCount;
    clone->analysisCount = this->analysisCount;

    // Освобождаем память, выделенную в конструкторе
    delete[] clone->topSales;
    delete[] clone->analysis;

    // ПОВЕРХНОСТНОЕ КОПИРОВАНИЕ: копируем указатели
    clone->topSales = this->topSales;
    clone->analysis = this->analysis;

    return clone;
}

// Метод для глубокого клонирования
SalesReport* SalesReport::deepClone() const {

    SalesReport* clone = new SalesReport();

    // Копируем простые поля
    clone->period = this->period + " (deep clone)";
    clone->money = this->money;
    clone->soldProducts = this->soldProducts;
    clone->topSalesCount = this->topSalesCount;
    clone->analysisCount = this->analysisCount;

    //  копируем содержимое массивов
    for (int i = 0; i < topSalesCount; i++) {
        clone->topSales[i] = this->topSales[i];
    }

    for (int i = 0; i < analysisCount; i++) {
        clone->analysis[i] = this->analysis[i];
    }

    return clone;
}

// Вывод информации о отчете
void SalesReport::displayReportInfo() const {
    cout << "\n=== ИНФОРМАЦИЯ ОБ ОТЧЕТЕ ===" << endl;
    cout << "Период: " << period << endl;
    cout << "Выручка: " << money << " руб." << endl;
    cout << "Проданные товары: " << soldProducts << " шт." << endl;
    cout << "Топ товаров: " << topSalesCount << endl;
    cout << "Записей анализа: " << analysisCount << endl;
}


// Модификация для демонстрации
void SalesReport::modifyReportForDemo() {
    cout << "\n=== МОДИФИКАЦИЯ ОТЧЕТА ДЛЯ ДЕМОНСТРАЦИИ ===" << endl;

    if (topSalesCount > 0) {
        topSales[0] = "МОДИФИЦИРОВАННЫЙ ТОВАР";
        cout << "Первый товар в топе изменен на: " << topSales[0] << endl;
    }

    money += 1000.0;
    cout << "Выручка увеличена на 1000 руб." << endl;
}

// Выделение памяти для массивов
void SalesReport::allocateMemory() {
    topSales = new std::string[MAX_TOP_PRODUCTS];
    analysis = new std::string[MAX_ANALYSIS_ITEMS];
}

// Освобождение памяти
void SalesReport::deallocateMemory() {
    delete[] topSales;
    delete[] analysis;
    topSales = nullptr;
    analysis = nullptr;
}

// Копирование массивов (глубокое копирование)
void SalesReport::copyArrays(const SalesReport& other) {
    // Копируем топ товаров
    for (int i = 0; i < other.topSalesCount; i++) {
        topSales[i] = other.topSales[i];
    }

    // Копируем анализ
    for (int i = 0; i < other.analysisCount; i++) {
        analysis[i] = other.analysis[i];
    }
}

// Поверхностное копирование массивов (только указатели)
void SalesReport::shallowCopyArrays(const SalesReport& other) {

    topSales = other.topSales;
    analysis = other.analysis;
}


void SalesReport::GetReport() {
    cout << "\n=== ОФОРМЛЕНИЕ ОТЧЕТА ПО ПРОДАЖАМ ===" << endl;
    cout << "Период отчета: " << period << endl;
    inputReportData();
    SaveToFile();
    cout << "Отчет успешно оформлен и сохранен!" << endl;
}

void SalesReport::RemakeReport() {
    cout << "\n=== ИЗМЕНЕНИЕ ОТЧЕТА ПО ПРОДАЖАМ ===" << endl;
    string filename = "sales_report_" + period + ".txt";

    if (!LoadFromFile(filename)) {
        cout << "Отчет за период '" << period << "' не найден!" << endl;
        return;
    }

    cout << "Текущие данные отчета:" << endl;
    DisplayReport();
    cout << "\nВведите новые данные:" << endl;
    inputReportData();
    SaveToFile();
    cout << "Отчет успешно изменен!" << endl;
}

void SalesReport::inputReportData() {
    // Ввод выручки
    while (true) {
        cout << "Введите выручку: ";
        cin >> money;
        if (cin.fail() || money < 0) {
            cin.clear();
            cin.ignore(numeric_limits<streamsize>::max(), '\n');
            cout << "Ошибка! Введите положительное число." << endl;
            continue;
        }
        break;
    }

    // Ввод количества проданных товаров
    while (true) {
        cout << "Введите количество проданных товаров: ";
        cin >> soldProducts;
        if (cin.fail() || soldProducts < 0) {
            cin.clear();
            cin.ignore(numeric_limits<streamsize>::max(), '\n');
            cout << "Ошибка! Введите целое положительное число." << endl;
            continue;
        }
        break;
    }

    cin.ignore(numeric_limits<streamsize>::max(), '\n');

    // Ввод топа товаров
    cout << "\n=== ТОП ТОВАРОВ ПО ПРОДАЖАМ ===" << endl;
    topSalesCount = 0;

    for (int i = 0; i < MAX_TOP_PRODUCTS; i++) {
        string product;
        cout << "Введите товар №" << (i + 1) << " (или 'stop' для завершения): ";
        getline(cin, product);

        if (product == "stop" || product.empty()) {
            break;
        }

        topSales[topSalesCount++] = product;
    }

    // Ввод аналитики
    cout << "\n=== АНАЛИТИКА ПРОДАЖ ===" << endl;
    analysisCount = 0;

    for (int i = 0; i < MAX_ANALYSIS_ITEMS; i++) {
        string analysisItem;
        cout << "Введите пункт аналитики №" << (i + 1) << " (или 'stop' для завершения): ";
        getline(cin, analysisItem);

        if (analysisItem == "stop" || analysisItem.empty()) {
            break;
        }

        analysis[analysisCount++] = analysisItem;
    }
}

void SalesReport::SaveToFile() {
    string filename = "sales_report_" + period + ".txt";
    ofstream file(filename);

    if (file.is_open()) {
        file << period << endl;
        file << money << endl;
        file << soldProducts << endl;
        file << topSalesCount << endl;

        for (int i = 0; i < topSalesCount; i++) {
            file << topSales[i] << endl;
        }

        file << analysisCount << endl;

        for (int i = 0; i < analysisCount; i++) {
            file << analysis[i] << endl;
        }

        file.close();
    }
}

bool SalesReport::LoadFromFile(const std::string& filename) {
    ifstream file(filename);

    if (!file.is_open()) {
        return false;
    }

    getline(file, period);
    file >> money;
    file >> soldProducts;
    file >> topSalesCount;
    file.ignore();

    for (int i = 0; i < topSalesCount; i++) {
        getline(file, topSales[i]);
    }

    file >> analysisCount;
    file.ignore();

    for (int i = 0; i < analysisCount; i++) {
        getline(file, analysis[i]);
    }

    file.close();
    return true;
}

void SalesReport::DisplayReport() const {
    cout << "\n=== ОТЧЕТ ПО ПРОДАЖАМ ===" << endl;
    cout << "Период: " << period << endl;
    cout << "Выручка: " << fixed << setprecision(2) << money << " руб." << endl;
    cout << "Проданные товары: " << soldProducts << " шт." << endl;

    if (topSalesCount > 0) {
        cout << "\nТоп товаров по продажам:" << endl;
        for (int i = 0; i < topSalesCount; i++) {
            cout << (i + 1) << ". " << topSales[i] << endl;
        }
    }

    if (analysisCount > 0) {
        cout << "\nАналитика продаж:" << endl;
        for (int i = 0; i < analysisCount; i++) {
            cout << "- " << analysis[i] << endl;
        }
    }

    cout << "==========================" << endl;
}

std::string SalesReport::getCurrentDate() {
    time_t now = time(0);

#ifdef _WIN32
    struct tm timeinfo;
    localtime_s(&timeinfo, &now);
    char buffer[80];
    strftime(buffer, 80, "%d.%m.%Y", &timeinfo);
#else
    tm* localTime = localtime(&now);
    char buffer[80];
    strftime(buffer, 80, "%d.%m.%Y", localTime);
#endif

    return string(buffer);
}